### PR TITLE
Highlight seek line using different background color

### LINKED ISSUE #9754: Feature request: Line highlight on seek
Currently, while seeking in Visual mode, r2 highlights only the offset. It would be nice to have an option that allows one to highlight the whole line while seeking. Visually, this highlight will stretch horizonally, and include all the columns including opcodes, comments etc. The highlight should adjust if `scr.breaklines` is enabled and work for multiline comments also. 

The cyan line in the sreenshot indicates how the highlight should look.

image

### PR DESCRIPTION
**Pull request by kazarmy**
To fix #9754.

### CODE REVIEW

**Comment by radare** [comment]
do you have a screenshot?

File: libr/core/disasm.c [hunk 1]
```diff
@@ -4583,6 +4585,17 @@ R_API int r_core_print_disasm(RPrint *p, RCore *core, ut64 addr, ut8 *buf, int l
 				r_cons_printf ("%s%s%s; --------------------------------------\n",
 					COLOR (ds, color_flow), ds->line, COLOR_RESET (ds));
 			}
+		}
+		if (ds->show_seekline && ds->vat == core->prompt_offset) {
+			int pad_len = 2 + R_MAX (0, ds->line ? strlen (ds->line) - 2 : 0);
+			int line_len = R_MAX (0, ds->interactive ? r_cons_singleton ()->columns : 80) - pad_len;
+			r_cons_memset (' ', pad_len);
+			r_cons_print (COLOR_CONST (ds, CYAN));
```
Comment (radare) [inline libr/core/disasm.c]: honor scr.color and dont hardcode the color, use the theme palette for this

**Comment by ret2libc** [comment]
I personally don't see a lot of needs for this feature. It's a line always after the first line, so it doesn't seem to help a lot. Moreover, it breaks the lines on the left and some times the seekline starts at a different  point of the screen.
screen shot 2018-04-10 at 20 42 07
screen shot 2018-04-10 at 20 42 17

**Comment by radare** [comment]
I was thinking in highlighting thw background of the selected line with another color from the palette

**Comment by XVilka** [comment]
Yes, setting the background can save the precious vertical space.

File: libr/core/disasm.c (hunk 1 above)
Comment (kazarmy) [reply libr/core/disasm.c]: can it have its own color in the theme palette?

**Comment by radare** [comment]
of course

**Comment by kazarmy** [comment]
> I was thinking in highlighting thw background of the selected line with another color from the palette

I am almost sure that there will be contrast problems with some foreground-background color combinations.

> It's a line always after the first line, so it doesn't seem to help a lot.

@ret2libc It does make the first line stand out more. Might help tired eyes.

> Moreover, it breaks the lines on the left ...

By design to reduce clutter.

> ... and some times the seekline starts at a different point of the screen.

This is probably a bug due to the combination of multibyte UTF8 chars and use of `strlen`. This appears fixable.

> I personally don't see a lot of needs for this feature.

I didn't code this feature for you (and for me neither tbh). It's disabled by default.

**Comment by radare** [comment]
> On 11 Apr 2018, at 16:34, Khairul Azhar Kasmiran <notifications@github.com> wrote:
> 
> I was thinking in highlighting thw background of the selected line with another color from the palette
> 
> I am almost sure that there will be contrast problems with some foreground-background color combinations.
> 
me too. but i will only show background when no bg is set by the theme. so it should work “out of the box"
> It's a line always after the first line, so it doesn't seem to help a lot.
> 
> @ret2libc <https://github.com/ret2libc> It does make the first line stand out more. Might help tired eyes.
> 
> 

i think bgcolor is the way
> Moreover, it breaks the lines on the left ...
> 
> By design to reduce clutter.
> 
> ... and some times the seekline starts at a different point of the screen.
> 
> This is probably a bug due to the combination of multibyte UTF8 chars and use of strlen. This appears fixable.
> 
> I personally don't see a lot of needs for this feature.
> 
> I didn't code this feature for you (and for me neither tbh). It's disabled by default.
> 
> 
:P

**Comment by kazarmy** [comment]
Since @securisec wanted a line, I think it's best to just close this and #9754.

**Comment by kazarmy** [comment]
I think highlighting the seek line would be more useful if it's actually halfway down the screen, but that might probably require fully solving the disassembly-in-reverse problem first.

**Comment by securisec** [comment]
@kazarmy why did you close both these issues if they are not resolved?

**Comment by kazarmy** [comment]
@securisec Do you accept the highlighting of the seek line by background color instead of a line under it? If not, or if there is no response within a few days, then I should close this PR though I'll leave the issue open.

**Comment by securisec** [comment]
@kazarmy Would you happen to have a screenshot of how it will look? Thanks

**Comment by kazarmy** [comment]
> @kazarmy Would you happen to have a screenshot of how it will look? Thanks

seekline_highlight

Well, if scr.color=3 is used, the colors are actually not bad. A code complication here is that there are a lot of places that use Color_RESET instead of the proper Color_NOBGRESET (i.e. reset everything except the background color).

**Comment by securisec** [comment]
Thanks @kazarmy . Based on the screenshot, this is exactly what i was hoping for in my feature request

**Comment by XVilka** [comment]
Awesome, thank you!

**Comment by radare** [comment]
Beautiful

**Comment by d4em0n** [comment]
how to enabled this feature ?

**Comment by kazarmy** [comment]
`e asm.highlight = $$` to highlight current seek
`e asm.highlight = eip` to highlight current location of eip
etc.

**Pull request merged**